### PR TITLE
fix oneOf with type => [array]

### PR DESCRIPTION
**Pull request by karenetheridge**
### Summary
fixes more edge cases with oneOf

### Motivation
Many of the new tests I added will fail without the accompanying code changes.

### CODE REVIEW

File: lib/JSON/Validator.pm [hunk 1]
```diff
@@ -573,69 +573,85 @@ sub _validate {
...
+
+  return prefix_errors allOf => @errors;
 }
 
 sub _validate_any_of {
-  my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@e, @errors, @expected);
+  my ($self, $data, $path, $rules, $prefix) = @_;
+  $prefix //= 1;
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: I don't like this `$prefix` variable. It's a bit worse than my `$expected` variable in `_validate_type_number()` 😥 Can we make prefix_errors() smarter instead?

File: lib/JSON/Validator.pm [hunk 2]
```diff
@@ -573,69 +573,85 @@ sub _validate {
...
+  my (@e, @errors);
 
   my $i = 0;
   for my $rule (@$rules) {
     @e = $self->_validate($_[1], $path, $rule);
+
+    # TODO: the spec might not think that short-circuiting is ok here
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: What do you mean with this comment? Why would you use CPU cycles checking for more errors, if it already fulfils anyOf?

File: lib/JSON/Validator.pm [hunk 3]
```diff
@@ -573,69 +573,85 @@ sub _validate {
...
+  if (grep $_->[1]->details->[1] ne 'type', @errors) {
+    return $prefix ? prefix_errors(anyOf => @errors) : (map $_->[1], @errors);
+  }
+
+  my $want_types = join('/', uniq map $_->[1]->details->[0], @errors);
+  return E $path,
+    [
+    ($prefix ? (anyOf => type => $want_types) : ($want_types => 'type')),
+    $errors[-1][1]->details->[2]
+    ];
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: Pretty sure this can be made short as well. See my comment above about "This is very difficult to read". (I'm thinking "extract `@details` first) This goes for `_validate_one_of()` below as well.

File: t/issue-103-one-of.t [hunk 4]
```diff
@@ -5,7 +5,8 @@ validate_ok {who_id => 'WHO', expire => '2018-01-01', amount => 1000,
   desc => 'foo'}, 'data://main/example.json',
   E('/sym',      '/oneOf/0/allOf/0/allOf/0 Missing property.'),
   E('/template', '/oneOf/0/allOf/2 Missing property.'),
-  E('/sym',      '/oneOf/1/allOf/0 Missing property.');
+  E('/sym',      '/oneOf/1/allOf/0 Missing property.'),
+  E('/',         '/oneOf/2 Expected string - got object.');
```
Comment (jhthorsen) [inline t/issue-103-one-of.t]: 👍

File: t/jv-allof.t [hunk 5]
```diff
@@ -39,4 +39,11 @@ validate_ok 'he110th3re', $schema,
   E('/', '/oneOf/1 String is too long: 10/4.'),
   E('/', '/oneOf/1 String does not match ^[a-z]+$.');
 
+
```
Comment (jhthorsen) [inline t/jv-allof.t]: Extra newline.

File: t/jv-allof.t [hunk 6]
```diff
@@ -39,4 +39,11 @@ validate_ok 'he110th3re', $schema,
...
+validate_ok 'hello', {type => ['integer', 'boolean']},
+  E('/', 'Expected integer/boolean - got string.');
```
Comment (jhthorsen) [inline t/jv-allof.t]: I think the message should have the types sorted, despite the order in `type`. (See #198)

File: t/jv-anyof.t [hunk 7]
```diff
@@ -63,4 +67,15 @@ validate_ok(
   E('/b', '/anyOf/1 Missing property.'),
 );
 
+validate_ok 'hello', {type => ['integer', 'string'], enum => [123, 'HELLO']},
+  E('/', 'Expected integer - got string.'),
```
Comment (jhthorsen) [inline t/jv-anyof.t]: This doesn't sounds right. We got a valid type, didn't we..?

File: t/jv-oneof.t [hunk 8]
```diff
@@ -23,29 +23,50 @@ $schema = {oneOf => [{type => 'object'}, {type => 'string', multipleOf => 3}]};
 validate_ok 13, $schema, E('/', '/oneOf Expected object/string - got number.');
 
 $schema = {oneOf => [{type => 'object'}, {type => 'number', multipleOf => 3}]};
-validate_ok 13, $schema, E('/', '/oneOf/1 Not multiple of 3.');
+validate_ok 13, $schema, E('/', '/oneOf/0 Expected object - got number.'),
+  E('/', '/oneOf/1 Not multiple of 3.');
 
 # Alternative oneOf
-# http://json-schema.org/latest/json-schema-validation.html#anchor79
+# https://json-schema.org/draft-07/json-schema-validation.html#rfc.section.7
```
Comment (jhthorsen) [inline t/jv-oneof.t]: 👍

File: lib/JSON/Validator.pm (hunk 2 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: The draft 2019-09 spec has a few things to say about verifying all branches of a conditional even if some parts already gave a sufficient answer to report up to the next level.  I didn't mean to commit that comment; it was a note to myself to go back and read the spec more carefully.

File: t/jv-allof.t (hunk 6 above)
Comment (karenetheridge) [reply t/jv-allof.t]: I disagree!  Any time there is a defined order in the schema (an arrayref, vs a hashref) I think we should keep the order as presented by the user.

File: t/jv-oneof.t (hunk 8 above)
Comment (karenetheridge) [reply t/jv-oneof.t]: It took a bit of sleuthing to figure out what part of the spec was being referenced there. I only figured it out when I broken the 'format' tests and found the bit in the spec explaining how it worked differently than everything else -- I had altered `schema_type` to infer `string` from `format`, and then found the tests that broke as a consequence.

**Comment by karenetheridge** [comment]
I have created a separate PR for the first change.  That commit is still contained in this PR as the code change is also needed here, but it will rebase away when it is merged to master.

File: lib/JSON/Validator.pm (hunk 1 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: I don't either, but it was late when I got to this point and I kind of gave up. :(  The key here is indicating to _validate_any_of whether it should actually be prefixing a path to the errors it got, or not -- when called via `type => [...]` there isn't any literal anyOf rule from the user, so we should just report the error(s) as-is without a prefix.

**Comment by karenetheridge** [comment]
`@expected` is removed in the _validate_(all|any|one)_of subs because it doesn't actually help anything -- the `_validate` call already did the type check and returned the proper error, and also `schema_type` can't return a useful result in the cases where `type => ..` is a list of types.

We don't actually need to do anything extra at all at the end of these subs if we don't care about joining all the `type` errors into one -- the code gets much simpler if instead of:

    E('/', '/anyOf/1 Expected integer/boolean - got string.');

we can return:

    E('/', '/anyOf/1 Expected integer - got string.');
    E('/', '/anyOf/1 Expected boolean - got string.');

File: t/jv-allof.t (hunk 6 above)
Comment (jhthorsen) [reply t/jv-allof.t]: Most people don’t care about order (based on experience working with people) when they write their schemas, so we don’t either when creating the error messages.

File: lib/JSON/Validator.pm [hunk 9]
```diff
@@ -519,9 +519,8 @@ sub _validate {
   # Test base schema before allOf, anyOf or oneOf
   if (ref $type eq 'ARRAY') {
     push @{$self->{temp_schema}}, [map { +{%$schema, type => $_} } @$type];
-    push @errors,
-      $self->_validate_any_of($to_json ? $$to_json : $_[1],
-      $path, $self->{temp_schema}[-1]);
+    return $self->_validate_any_of_types($to_json ? $$to_json : $_[1],
+      $path, $self->{temp_schema}[-1], 0);
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: Extra ", 0" at the end.

File: lib/JSON/Validator.pm [hunk 10]
```diff
@@ -574,69 +573,101 @@ sub _validate {
 
 sub _validate_all_of {
   my ($self, $data, $path, $rules) = @_;
-  my $type = data_type $data, $rules;
-  my (@errors, @expected);
+  my (@errors);
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: No need for parenthesis.

File: lib/JSON/Validator.pm [hunk 11]
```diff
@@ -574,69 +573,101 @@ sub _validate {
...
+    [
+    allOf => type => join('/', uniq map $_->[1]->details->[0], @errors),
+    $errors[-1][1]->details->[2]
+    ]
+    if @errors > 1 and not grep $_->[1]->details->[1] ne 'type', @errors;
+
+  return prefix_errors allOf => @errors;
+}
+
+sub _validate_any_of_types {
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: 👍

File: lib/JSON/Validator.pm [hunk 12]
```diff
@@ -574,69 +573,101 @@ sub _validate {
...
+  my ($self, $data, $path, $rules) = @_;
+  my @errors;
+
+  for my $rule (@$rules) {
+    my @e = $self->_validate($_[1], $path, $rule);
+    return unless @e;
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: I think you can merge these two lines: `return unless my @e = $self->_validate($_[1], $path, $rule);`, or maybe this even:

```
push @errors, $self->_validate($_[1], $path, $rule);
return unless @errors;
```

File: lib/JSON/Validator.pm [hunk 13]
```diff
@@ -574,69 +573,101 @@ sub _validate {
...
+    push @errors, @e;
+  }
+
+  if (my @e = grep $_->details->[1] ne 'type', @errors) {
+    return @e;
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: I think this is not what you want. Why filter out the "type" errors? See my comment below for a fixed version.

File: lib/JSON/Validator.pm [hunk 14]
```diff
@@ -574,69 +573,101 @@ sub _validate {
...
+  }
+
+  # combine all 'type' errors together
+  my $want_types = join('/', uniq map $_->details->[0], @errors);
+  return E $path, [$want_types => 'type', $errors[-1]->details->[2]];
```
Comment (jhthorsen) [inline lib/JSON/Validator.pm]: I would like to use `@details` in the code above:

```
  my @details = grep { $_->[1] eq 'type' } map { $_->details } @errors;
  return @errors unless @details == @errors;

  my $want_types = join '/', map { $_->[0] } @details;
  return E $path, [$want_types => type => $details[-1][2]];
```

File: t/jv-allof.t (hunk 6 above)
Comment (karenetheridge) [reply t/jv-allof.t]: I have not changed the sorting in this PR. Let's look at that later.

File: t/jv-anyof.t (hunk 7 above)
Comment (karenetheridge) [reply t/jv-anyof.t]: fixed, by splitting apart the anyOf logic into two subs, and the array-type-specific version will disregard mismatched type errors if there are more specific errors we can report instead.

File: lib/JSON/Validator.pm (hunk 9 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: let's look at combining them afterwards, to keep the changes here simpler. I don't think it will help any (the subs are all pretty short now, but subtly different in how they manage the loop - one of them short-circuits and the others don't; one keeps track of passing tests and the others don't..), so combining them might make readability worse.

File: lib/JSON/Validator.pm (hunk 10 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: I removed most of them but missed this one.

File: lib/JSON/Validator.pm (hunk 12 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: agreed.

File: lib/JSON/Validator.pm (hunk 13 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: Filtering out the type errors resolves the complaint you had about the 'enum' tests, where we were reporting a mismatched type even though the other choice of type did match. (t/jv-anyof.t line 70)

File: lib/JSON/Validator.pm (hunk 14 above)
Comment (karenetheridge) [reply lib/JSON/Validator.pm]: unfortunately that breaks some tests.  I don't see a good way of shortening the code without changing the logic.

**Comment by jhthorsen** [comment]
Thanks for all the changes 👍
